### DISCUSSION TITLE
Traduzioni di oggetti in frontend

### DISCUSSION THREAD

**Discussion by edoardocavazza** [chialab/bedita-frontend-kit#46] [Ideas]
## Desiderata

* Possibilità di mostrare l'oggetto tradotto secondo il linguaggio richiesto dall'utente
* Possibilità di indicare le traduzioni disponibili dell'oggetto

## Cose da sapere

* BEdita (core) consente di creare traduzioni per gli oggetti. Ogni oggetto dispone di una proprietà che la lingua originale, e di possibili traduzioni per le lingue configurate.
* Le traduzioni non sono relazioni, ma semplici associazioni Cake. Tuttavia, possono essere scaricate tramite la `RelatedObjectsAction` o tramite l'include via API/loader.
* La `GetObjectAction`. che usiamo per recuperare gli oggetti, automaticamente include la traduzione richiesta tramite il parametro `lang`. NOTA BENE: include la traduzione come associazione (`$object->translations`), ma non traduce i campi dell'oggetto originale.

## Analisi

Valutiamo in questa sezione tutti gli aspetti richiesti per raggiungere i desiderata.

### Routing

Dobbiamo per cominciare introdurre le regole di routing per caricare l'oggetto tradotto. Generalmente, usiamo `/lang/:lang/path/to/object`

Per ogni lingua configurata possiamo quindi prefissare il routing di default passando il parametro `_locale` che ci arriva dall'url.

#### Middleware

Il `LocaleMiddleware` (da introdurre in Frontend Kit) riceverà il parametro `_locale` dalla richiesta e lo imposterà come locale dell'app.

Questo impatta il loader di oggetti del Frontend Kit, le traduzioni che vengono generate tramite `__` e quelle di interfaccia scritte usando i file PO.

Da valutare anche il middleware di [BEdita/I18n](https://github.com/bedita/i18n/blob/master/src/Middleware/I18nMiddleware.php) per capire se può fare a caso nostro anche in caso di overkill.

#### Redirects

Dobbiamo capire come e se gestire i redirect per linguaggi espressi con pattern diversi, ad esempio (nel caso il linguaggio di default dell'app sia l'italiano):

`/home` (200)
`/lang/it/home` -> `/home` (302)
`/lang/ita/home` -> `/home` (o `/lang/it/home`?) (302)
`/lang/en/home` (200)
`/lang/eng/home` -> `/lang/en/home` (302)

In caso di linguaggio non configurato:
`/lang/ch/home` (404) oppure redirect a linguaggio principale `/home` (302)

#### ObjectRoute

Il metodo che si incarica di trovare la rotta giusta per un oggetto quando richiamato tramite `/objects/<uname>` deve tener conto del parametro `_locale` che arriva dalla richiesta.

### Attraente magia oscura

Quando, tramite routing, l'utente richiede un oggetto per uno specifico linguaggio il FrontendKit si incarica di scaricare la traduzione richiesta e sostituire i campi tradotti. Per sapere quali campi sono stati tradotti, sarà comunque possibile accedere alla proprietà `_originalFields` dell'oggetto.

**Pro**
* Just works™️
* Non richiede grandi refactor se le traduzioni vengono introdotte in un secondo (o terzo/quarto/quinto) momento

**Cons**
* Modificare la entity originale è sempre molto dubbio
* In caso di frontend che permetto l'editing in pagina dei contenuti e delle traduzioni* bisogna prestare attenzione che l'oggetto tradotto non sovrascriva quello in lingua originale

> \* Caso ipotetico. Non mi risulta sia mai stato sviluppato un frontend del genere.

### Noiosa magia chiara

Così come già accade, quando l'utente richiede un oggetto per uno specifico linguaggio il FrontendKit si incarica di scaricare la traduzione e di includerla come associazione nell'oggetto. In vista, potremo decidere se mostrare il titolo tramite `{{ object.title|raw }}` oppure se usare un helper per accedere alla traduzione, se disponibile: `{{ I18n.field(object, 'title', _currentLocale) }}`.

**Pro**
* Non si modifica l'oggetto originale
* L'utilizzo della traduzione è esplicito
* È un pattern già utilizzato dai partner

**Cons**
* Sicuramente più verboso
* Richiede refactor di tutte le viste se non previsto da inizio progetto*

> \* cosa che, IMHO, vale comunque prendere l'abitudine di fare, anche per le label di interfaccia, e anche se scoccia. Ma demanderei ad altra discussione

**Comment by qwerg** [comment]
Ovviamente **Attraente magia oscura**
Un normale flusso redazionale funziona così:

- esce il sito in italiano
- nel tempo (anche moolto tempo dopo) la redazione comincia a tradurre alcuni pezzi. Appare l'indicazione all'interno del singolo oggetto che esistono N traduzioni linkate). Non c'è ancora uno switch generale tra lingue.
- quando la % di traduzioni è accettabile si attiva lo switch generale del sito per le traduzioni con buona percentuale, ex ` ita | eng | deu | fra`
La scelta della lingua in base alle impostazioni del browser/SO, se la si vuole, dovrebbe attivarsi soltanto quando è già presente lo switch generale.

La magia oscura fa sì che non si debba fare un codice verbosissimo per un sito per il quale forse mai apparirà una traduzione.
